File: server/judge/preparators/jsPreparator.ts
import Preparator from ".";
import { Problem } from "../../problems";
import { TEMPLATE } from "../codeTemplate";
import { TestCase } from "../../problems";

const jsPreparator: Preparator = {
  // This function creates the final code that'll be executed for Javascript
  prepare(userCode: string, problem: Problem, testCase: TestCase): string {
    let codeToExecute = TEMPLATE.replace("{imports}", "");

    let variableDeclarations: string[] = [];
    Object.keys(problem.input).forEach((key) => {
      const inputType = problem.input[key].type;
      if (inputType === "array" || inputType === "object") {
        variableDeclarations.push(`let ${key} = ${JSON.stringify(testCase.input[key])};`);
      } else if (inputType === "string") {
        variableDeclarations.push(`let ${key} = "${testCase.input[key]}";`);
      } else {
        variableDeclarations.push(`let ${key} = ${testCase.input[key]};`);
      }
    });
    codeToExecute = codeToExecute.replace(
      "{var_declarations}",
      variableDeclarations.join("\n")
    );

    codeToExecute = codeToExecute.replace(
      "{user_code}",
      userCode
    );

    let functionCall = `${problem.functionName}(${Object.keys(problem.input).join(", ")});`;
    if (problem.executionMode === "return") {
      functionCall = `let result = ${functionCall}`;
    }
    codeToExecute = codeToExecute.replace(
      "{function_call}",
      functionCall
    );

    let printResult = "";
    if (problem.executionMode === "return") {
      printResult = `console.log(result);`;
    } else {
      const outputVariable = Object.keys(problem.input).filter(key => problem.input[key].output)[0];
      printResult = `console.log(${outputVariable});`;
    }

    codeToExecute = codeToExecute.replace(
      "{print_result}",
      printResult
    );

    return codeToExecute;
  }
};

export = jsPreparator;

File: server/judge/preparators/jsPreparator.go
package preparators

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/sriramr98/dsa_server/problems"
	"github.com/sriramr98/dsa_server/utils"
)

type JsPreparator struct{}

func (p JsPreparator) Prepare(userCode string, problem problems.Problem, testCase problems.TestCase) (string, error) {
	codeToExecute := strings.ReplaceAll(CODE_TEMPLATE, "{imports}", "")

	var variableDeclarations []string
	for inputKey, inputInfo := range problem.Input {
		var declaration string
		switch inputInfo.Type {
		case problems.ArrayType, problems.ObjectType:
			testCaseJSON, err := json.Marshal(testCase.Input[inputKey])
			if err != nil {
				return "", err
			}
			declaration = fmt.Sprintf("let %s = %s;", inputKey, string(testCaseJSON))
		case problems.StringType:
			declaration = fmt.Sprintf("let %s = \"%s\";", inputKey, testCase.Input[inputKey])
		default:
			declaration = fmt.Sprintf("let %s = %v;", inputKey, testCase.Input[inputKey])
		}

		variableDeclarations = append(variableDeclarations, declaration)
	}

	varDeclarations := strings.Join(variableDeclarations, "\n")
	codeToExecute = strings.ReplaceAll(codeToExecute, "{var_declarations}", varDeclarations)

	codeToExecute = strings.ReplaceAll(codeToExecute, "{user_code}", userCode)

	funcArgs := utils.JoinStringSlice(utils.MapKeysSorted(problem.Input), ", ")
	functionCall := fmt.Sprintf("%s(%s)", problem.FunctionName, funcArgs)

	if problem.ExecutionMode == problems.ReturnMode {
		functionCall = fmt.Sprintf("let result = %s", functionCall)
	}

	codeToExecute = strings.ReplaceAll(codeToExecute, "{function_call}", functionCall)

	var printResult string
	if problem.ExecutionMode == problems.ReturnMode {
		if problem.Output.Type == problems.ArrayType || problem.Output.Type == problems.ObjectType {
			printResult = "console.log(JSON.stringify(result));"
		} else {
			printResult = "console.log(result);"
		}
	} else {
		var outputVariable string
		for inputKey, inputConf := range problem.Input {
			if inputConf.Output {
				outputVariable = inputKey
				break
			}
		}

		if outputVariable == "" {
			return "", errors.New("no output variable found for InPlace return type")
		}

		if problem.Input[outputVariable].Type == problems.ArrayType || problem.Input[outputVariable].Type == problems.ObjectType {
			printResult = fmt.Sprintf("console.log(JSON.stringify(%s));", outputVariable)
		} else {
			printResult = fmt.Sprintf("console.log(%s);", outputVariable)
		}
	}

	codeToExecute = strings.ReplaceAll(codeToExecute, "{print_result}", printResult)

	return codeToExecute, nil
}
